Add draft integration of glimmer-scoped-css

backspace opened the pull request:
This should not be merged, it’s only for discussion, as `glimmer-scoped-css` is not yet released. It works for me locally when I use a filesystem path in `package.json` that points to the result of `pnpm pack` in `glimmer-scoped-css/glimmer-scoped-css`. Although it crashes on every rebuild, unclear what that’s about, it needs more investigation, maybe just related to being installed from the filesystem:

```
file changed addon/components/header/index.gts
Build Error (WaitForTrees)

ENOTEMPTY: directory not empty, rmdir '/private/var/folders/34/1ldk02sx5ls7sdn1qvy5z2mr0000gn/T/embroider/c50e3e/packages/boxel-ui/lib/setup-ast-transforms/'
```

What is the future of BEM when we have scoped CSS, since it’s partially a nomenclature to prevent namespace collisions? For `Boxel::Header` I took this approach:
* removing the “block” class name in favour of using the top-level element name (`header`)
* keeping the “element” with the block name removed (`.content`)
* keeping the “modifier”s with the block name replaced with the element name (`header.--no-background`)

However, `Boxel::FieldContainer` (seems to have been partially renamed from `::Field`?) doesn’t have a top-level element with a stable element name. So for that I just left the styles as-is. Maybe that makes the most sense?

Another arbitrary choice I made: putting the `<style>` at the bottom of the `<template>`.

burieberry commented:
> Overall seems good! Wondering about a few things:
> 
> 1. How are we going to add the dropdown menu for the '...'?
>    ![image](https://user-images.githubusercontent.com/39579264/112136651-344daf00-8c0a-11eb-85e1-2e5574843767.png)
> 2. Button in the `LayoutContainer` header has odd focus states. We do need something to signal interactivity and this does do it, but this feels a bit odd. I think this needs some specification from design on the tabbing behaviour for a proper solution though.
>    ![image](https://user-images.githubusercontent.com/39579264/112137971-e8036e80-8c0b-11eb-96f4-b9804ddedf97.png)
> 3. If we change the word 'Layout', `LayoutContainer` can also be used for non-layout cards, so it feels a bit awkward if we end up making another one for non-layout cards which have the same affordances in the top right corner that allows a dynamic card name. Is it okay to allow people to pass the name of the card in via an argument?
>    ![image](https://user-images.githubusercontent.com/39579264/112138257-43356100-8c0c-11eb-9e9f-99c145a105f0.png)
>    ![image](https://user-images.githubusercontent.com/39579264/112139026-36fdd380-8c0d-11eb-9f36-3b6e6175e805.png)
> 4. There are a couple of states for cards. Wondering whether it is good to have a component that implements them here, or if it is best left to the actual usage via `CardContainer`.
>    ![image](https://user-images.githubusercontent.com/39579264/112140276-c657b680-8c0e-11eb-9281-fa1d63231218.png)
>    ![image](https://user-images.githubusercontent.com/39579264/112140294-cd7ec480-8c0e-11eb-97fd-5e60463b62b9.png)
>    ![image](https://user-images.githubusercontent.com/39579264/112140354-dcfe0d80-8c0e-11eb-9e08-3e41460d9363.png)
>    ![image](https://user-images.githubusercontent.com/39579264/112139026-36fdd380-8c0d-11eb-9f36-3b6e6175e805.png)

Here's what I'm thinking:

1- We can use the Boxel DropdownButton component for the menu. I haven't implemented it in this PR.

2- Yes, the focus tab styling is ugly and we should do a discuss with the design team. So far we're only using that in Card Space though, so maybe it can wait until we go back to working on Card Space.

3 and 4- Good point. I'm thinking of those cards as "embedded layout container"s. So we would use the current LayoutContainer for the outer container, and make an embedded version for the other cards. The view mode for those cards look different (they keep their boundaries and rounded corners, unlike the non-embedded one.) and they have the different states you've mentioned. More detailed explanation [here](https://docs.google.com/document/d/1xHLsur-dnGUbhLuqMU9kQHLr-TzQmuO6clTUSMYl6OI/edit?usp=sharing), even though I haven't updated it in a while. Will go back to implementing this during Card Space work.

I'll update the PR to add these as To-Do items.

Aierie reviewed (approved):
Cool, I think keeping changes minimal till we use the components is good then!

burieberry commented:
@lukemelia experiencing random test fail again, do you know why ember-beta is failing?

lukemelia commented:
> @lukemelia experiencing random test fail again, do you know why ember-beta is failing?

@burieberry I think we need to upgrade to embroider 0.37.0 to get https://github.com/embroider-build/embroider/commit/5bd5507cda84e28ff37df2a14040789a71e6276c /cc @Aierie

Aierie commented:
@lukemelia Upgrading to embroider 0.37.0 creates new errors that cause tests to fail both on our current version of ember, and `ember-beta`. Might need some help with debugging them

lukemelia commented:
@Aierie phone a friend

backspace committed:
Update Embroider

backspace committed:
Add glimmer-scoped-css

This is unreleased so of course this filesystem reference
should not be merged.

backspace committed:
Add possible example of Boxel scoped styles

backspace committed:
Add styles for FieldContainer

ef4 commented on packages/boxel-ui/addon/components/header/index.gts:
```diff
@@ -20,11 +20,11 @@ interface Signature {
-      boxel-header--no-background=@noBackground
-      boxel-header--highlighted=@isHighlighted
-      boxel-header--large=(or (bool @title) (eq @size 'large'))
+    {{!-- Can the first argument be optional? --}}
```
It looks like you could get that effect by skipping `cn` and directly calling the library that it calls.

ef4 commented on packages/boxel-ui/addon/components/field-container/index.gts:
```diff
@@ -51,6 +51,67 @@ const FieldContainer: TemplateOnlyComponent<Signature> = <template>
   {{/let}}
+
+  <style>
+    .boxel-field {
```
I agree that the variable element name means it's better to keep using this class.

ef4 commented on packages/boxel-ui/addon/components/field-container/index.gts:
```diff
@@ -51,6 +51,67 @@ const FieldContainer: TemplateOnlyComponent<Signature> = <template>
+      gap: var(--boxel-sp-xs) 0;
+    }
+
+    .boxel-field--vertical {
```
But there's no reason to keep namespacing all these other selectors with a `.boxel-field` prefix.

There's also no reason to keep the double dashes around. This could be just `.vertical`.

ef4 reviewed (commented):
Yes, I think BEM goes away completely with the switch to scoped CSS.

Once you have scoping, you can pick your class names the same as any local variable name. Something meaningful with only enough description to distinguish it from other things in the local scope.

backspace commented:
Thanks for the commentary, I’ll close this now but it’ll be useful when we truly integrate 😀
